### ISSUE TITLE
Job success/completion policy

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes/kubernetes#117600]
### What would you like to be added?

An API to define when a Job (could be exclusive to Indexed) can be declared as succeeded. Examples of policies:
- Complete if Index 0 is successful
- Complete if x% of indexes are successful
- Complete if x indexes are successful

### Why is this needed?

- flux-operator: requires this behavior for MiniClusters (index 0 is the broker)
- kubeflow: some ML applications can terminate if a percentage of workers finish

/wg batch
/sig apps

**Comment by alculquicondor** [comment]
/triage needs-information

I would like to collect feedback from the community

**Comment by alculquicondor** [comment]
I think that comes from the same kubeflow requirement

**Comment by tenzen-y** [comment]
The kubeflow/training-operator has similar features in TFJob:

https://github.com/kubeflow/training-operator/blob/ddf372c8fe0d179674f3dc5e61fbd71246c32924/pkg/apis/kubeflow.org/v1/tensorflow_types.go#L77-L80

https://github.com/kubeflow/training-operator/blob/ddf372c8fe0d179674f3dc5e61fbd71246c32924/pkg/controller.v1/tensorflow/tfjob_controller.go#L505-L534

So, if this feature is implemented in batch/job, it is helpful to us.
If @alculquicondor does not work on this, I'd like to work on this. As a first step, need to submit a KEP?

**Comment by alculquicondor** [comment]
Thanks @tenzen-y. Yes, a KEP would be necessary.
However, maybe we can wait for others to provide more feedback. I'll mention this issue in tomorrow's meeting just in case.

**Comment by tenzen-y** [comment]
> However, maybe we can wait for others to provide more feedback. 

Sure, let's collect other feedback :)

> I'll mention this issue in tomorrow's meeting just in case.

Thanks!

**Comment by vsoch** [comment]
> Configurable success status. For example, allow users to specify that the job should be considered successful if 80% of workers finished successfully.

For some quick feedback, for our use case we _only_ care about index 0 of the IndexedJob. If index 0 completes, we are good to go. The rest of the pods are just workers in that cluster, and (with the current design we have) we keep them alive even with failures to start, because we retry. So the perfect design would be:

- When index 0 is completed, mark all the other workers as completed (in scope for this issue)
- Allow custom logic to bring up index 0 reliably first (out of scope for this issue but I thought I'd point it out!)

For the latter, @alculquicondor and I discussed the idea of a "child job" for another design problem (that is now resolved) and I wonder if it could be used in this separate context to solve the second bullet - have index 0 be a single job that is brought up as the broker, and then the workers brought up as a child job (IndexedJob) to it. When the broker job completes, it (should?) clean up the children. 

I think I'd rather have all of the logic in one Indexed Job, with custom actions for the broker (and wait for this issue to be resolved), but if you think this is an idea worth pursuing (that would handle this second design issue that the workers don't always complete cleanly) I could work on it next.

**Comment by ahg-g** [comment]
@vsoch as part of the JobSet effort (https://github.com/kubernetes-sigs/jobset), we are targeting use cases like the two you mentioned above. In a nutshell, it is an API to create and manage a group of Jobs as a unit. 

You can either create Jobs with different templates or many jobs of the same template. With this, one can easily employ the coordinator-worker pattern as two Jobs under JobSet. [Success policy](https://github.com/kubernetes-sigs/jobset/issues/82) and startup sequence are on the roadmap and should be simple to support.

**Comment by vsoch** [comment]
> You can either create Jobs with different templates or many jobs of the same template. With this, one can easily employ the coordinator-worker pattern as two Jobs under JobSet. https://github.com/kubernetes-sigs/jobset/issues/82 and startup sequence are on the roadmap and should be simple to support.

Love that! I don't know how this process works (and I'll likely learn as I watch here) but I'd love to try something out if/when it comes around. Please ping me anywhere I can be helpful or part of the discussion.

**Comment by kerthcet** [comment]
> Complete if Index 0 is successful

It seems index0 acts in a different mode with indexX, as mentioned above, it also requires to be brought up firstly, so I think it should be treated separately, like jobSet who provides different templates for jobs, or parent-child mode. Define the success/completion policy for this is somehow helpful but couldn't solve everything, do we have other cases?

I think we can support x/x% indexes success as a startup at the least.

**Comment by vsoch** [comment]
I can probably think of a few more cases for which a success/completion policy is important to know.

## Distributed Task Server

For the worker/broker case, any kind of distributed setup (e.g., tensorflow, pytorch, fireworks) will require a slightly different startup sequence for the main task manager vs. workers, and the completion policy might depend on the tool. As an example, for a worker pod receiving tasks, it might have a transient error and the pod fails. We'd want to restart the pod so the broker can re-launch the worker. If the broker fails we are in bigger trouble, and we probably want it to fail.

We could also spin this to be more interesting - having a completion policy that says "when a worker is done, clean up the pod" (freeing more resources on the cluster node) but allow the other pods to keep running. If you imagine a workflow tool submitting many JobSets (each with a broker and some number of workers) that are all sharing the same Kubernetes resources, this would help to fit more in. And then within one JobSet, we would have the opposite pattern to the use case I outlined above (workers complete when job is complete) where the broker might only exist to do some initial sharding, and then can be rendered complete when all the workers are done. 

## Dynamism

The above hints at dynamism, and this is something we are very interested in at the lab! E.g., it should be possible to scale up and down based on either the request of a worker (the broker) or if that's not possible, some overhead process that notices all of the resources for a JobSet are getting over-utilized (or under-utilized) and respond appropriately. If Kubernetes is hard to scale (and takes forever with the command line clients) maybe this is the way we can achieve something simpler - scaling the size of JobSets up and down based on user needs. For this case, I'd imagine that the creator would set a min and max that would be allowed for scaling. It would be super cool, and work well with workflow tools that run different sets of tasks that each need different container bases and resources.

Is that along the lines of what you are looking for? I've been testing a bunch of ML libraries, simulations, and workflows in the Flux Operator, and definitely have a lot of thoughts on the design (and can think more about mapping to this Batch API space). So far I've been using the hostname and/or job index environment variable for custom logic, and it's a lot of if/else statements in common templates, so I'm really looking forward to having JobSet!

**Comment by alculquicondor** [comment]
> It seems index0 acts in a different mode with indexX, as mentioned above, it also requires to be brought up firstly

As also @vsoch points out, we cannot really guarantee this. It is always possible that the index 0 comes up, then we create the rest of the indexes, but index 0 crashes in the meantime. For that reason, it's preferred that the application itself knows how to wait for the broker. Even OpenMPI works this way (not Intel MPI, sadly).

So I would be opposed to adding startup order within a Job.

Re: dynamism
https://github.com/kubernetes/enhancements/issues/3715 this is beta in 1.27. JobSet will capitalize on this.

However, while related, this thread is about Job, not JobSet. So the question is:

Is it useful to have the policy: "terminate job as succeeded if index 0 finishes"?
Another example of a job where index 0 is a broker is PyTorch, but I don't think it needs these semantics. @mimowo?

My question to @vsoch is: will you need this capability in Job, or will you migrate to JobSet?

**Comment by alculquicondor** [comment]
> We’d probably still use that design for the worker, but having reliably brought the broker up first, it would more regularly connect the first try

Why is that necessary? I think it's better if the workers know how to retry. Again, this is how OpenMPI works and it works pretty well.
Creating the workers at a deferred time is much slower, as we need to wait for k8s to: pull the image, setup the networking, start the container, wait for the actual broker server to be ready. So much time wasted that can be avoided if the workers now how to retry (which it sounds like they already know how to). This is a more robust solution in a distributed system such as kubernetes.

**Comment by vsoch** [comment]
This discussion had me think a bit and I think you’re right - it’s not that the workers should require that sleep/retry loop (which is a bit janky and prevents the job from completing when the worker exits) but flux start should have a wait flag that essentially does the same, but then allows the worker to exit (and the pod to exit and the job index to complete after). I’ll bring this up with the flux devs because it would solve this problem nicely I think.

**Comment by tenzen-y** [comment]
Through all the comments in this thread, I think we need to support the below two features for now:

- Complete if x% of indexes are successful
- Complete if x indexes are successful

This means we don't consider the `Complete if Index 0 is successful` policy mentioned in this issue description.

However, we need to consider the existing limitations and features, as @alculquicondor says:

> - NonIndexed Jobs have the semantics of "declare the job succeeded if any pod completes", when completions=nil. This is somewhat similar to one of the use cases listed. However, IndexedJobs currently restrict completions!=nil.
> - For use cases "x number of pods succeeded or x% number of pods succeeded", they could theoretically be served by a single API field https://github.com/kubernetes/apimachinery/blob/master/pkg/util/intstr/intstr.go#L172

https://github.com/kubernetes/kubernetes/issues/117600#issuecomment-1525855100

@alculquicondor Is this right?

**Comment by vsoch** [comment]
This won't help us at all, and it also won't help the general use case of a leader and followers or workers, but that's fine, we can continue with hacks.

**Comment by tenzen-y** [comment]
Uhm. It sounds like you have a point.

Actually, in the Distributed TF Training, I have a use case to want to care whether a Pod with index 0 succeeded.

https://github.com/kubeflow/training-operator/blob/ddf372c8fe0d179674f3dc5e61fbd71246c32924/pkg/controller.v1/tensorflow/tfjob_controller.go#LL507C4-L507C4

**Comment by alculquicondor** [comment]
The next question would be if it's enough to specialize index 0 only, or are there cases where you might need a range of indexes to complete?

Or... not specialize index 0 at all. You can still implement the semantics of "finish when index 0 completes" with:
"Complete if x indexes are successful", where `x=1`, if you are sure that no pod, other than index 0 has the ability to complete with success.

WDYT @vsoch ?

**Comment by vsoch** [comment]
> Or... not specialize index 0 at all. You can still implement the semantics of "finish when index 0 completes" with:
"Complete if x indexes are successful", where x=1, if you are sure that no pod, other than index 0, has the ability to complete with success.

yes I like that even better! Index 0 is not special, it really is an arbitrary choice, as is the specific set of pods you choose. E.g., you could imagine having 5 leaders in a set, indices 0-4, and you care about their completion. +1 from me for "Complete if x indices are successful where x is in the set..."

**Comment by alculquicondor** [comment]
"Complete if x indexes are successful" would be about any arbitrary index. Are you sure this is not enough?

We could add "complete the job if the set of x, y, z indexes succeed", but it's better to limit the API unless absolutely necessary.

**Comment by alculquicondor** [comment]
> you could imagine having 5 leaders in a set, indices 0-4, and you care about their completion

Right, but if only those indexes have the ability to have exitCode=0, then we can implement the rule as "any 5 indexes finish".

**Comment by vsoch** [comment]
> "Complete if x indexes are successful" would be about any arbitrary index. Are you sure this is not enough?

It is absolutely not enough - for my use case (or a general leader / workers setup) we only care about the leader(s). Even when the workers are completely done, the leader may be doing some finalization for assembling results, etc. 

> Right, but if only those indexes have the ability to have exitCode=0, then we can implement the rule as "any 5 indexes finish".

Any 5 indices would not work per the example above.

**Comment by tenzen-y** [comment]
> Through all the comments in this thread, I think we need to support the below two features for now:
> 
> * Complete if x% of indexes are successful
> * Complete if x indexes are successful
> 
> This means we don't consider the `Complete if Index 0 is successful` policy mentioned in this issue description.
> 
> However, we need to consider the existing limitations and features, as @alculquicondor says:
> 
> > * NonIndexed Jobs have the semantics of "declare the job succeeded if any pod completes", when completions=nil. This is somewhat similar to one of the use cases listed. However, IndexedJobs currently restrict completions!=nil.
> > * For use cases "x number of pods succeeded or x% number of pods succeeded", they could theoretically be served by a single API field https://github.com/kubernetes/apimachinery/blob/master/pkg/util/intstr/intstr.go#L172
> 
> [#117600 (comment)](https://github.com/kubernetes/kubernetes/issues/117600#issuecomment-1525855100)
> 
> @alculquicondor Is this right?

Let me summarize the semantics we want for now.

- Complete if a set of [x, y, z...] Indexs are successful
- Complete if x% of indexes are successful
- Complete if x indexes are successful

@vsoch @alculquicondor Does that look good?

@alculquicondor Through this issue, we have two concrete use cases, kubeflow and flax-operator.
So, I think we can start to write a KEP, right? Or should we collect more use cases?

**Comment by alculquicondor** [comment]
one more thing to consider... do we really need support for x%? Can the user just calculate the precise number on their own?

In Deployments, we allow "integer or percent" for fields like MaxSurge. This makes a lot of sense because the author of the Deployment might not know how many replicas the deployment will have, as it can autoscaled later.

Do we have the same justification for Jobs? elastic jobs perhaps?

> So, I think we can start to write a KEP, right? Or should we collect more use cases?

I think those are solid cases. @kubernetes/sig-apps-leads  any concerns before writing a KEP?

**Comment by tenzen-y** [comment]
> Do we have the same justification for Jobs? elastic jobs perhaps?

I agree. In general, the author of elastic jobs, such as elastic horovod, doesn't know the exact number of replicas in running.
So I think we should support `Complete if x% of indexes are successful`.

**Comment by vsoch** [comment]
We are hugely interested in the elasticity / dynamism use case! And although it doesn't fit my use case, I can imagine this might be rationale to allow a percentage. But I think that might be a separate thing from completion - e.g., for the Flux operator (or a leader/worker setup) you would want the design:

- the job is complete when the main leaders (0...n of N) are complete
- we can remove / add workers as we please (and their removal we don't care about)

I think the subtle clash here is that having a rule like "be done when % of workers are done" is making the underlying assumption that the responsibility for managing the workers is up to the indexed job (or similar Job). But actually in most of these leader / worker scenarios, it's the leader that is responsible for determining when a worker is done. And the indexed job just needs to ensure that the base pod (onto which a worker will be launched) is always ready / recreated when something happens, and only complete when some pre-determined set of leaders is finished.

So what we are really caring about is looking for leader completion, some subset of indices, and making sure the others keep running. So +1 from me for that list of 3 use cases.

**Comment by tenzen-y** [comment]
If @kubernetes/sig-apps-leads doesn't have any concerns about this feature. 
As a first step, I'd like to write a KEP in google doc.

@alculquicondor Or, would you like to take this?

**Comment by alculquicondor** [comment]
No, you can take it. I think it's better to write the KEP directly, otherwise comments might get lost in the google doc. A google doc might be more useful when we still have very abstract ideas of what we want to do. But here we have all the use cases listed already :)

**Comment by tenzen-y** [comment]
> No, you can take it. I think it's better to write the KEP directly, otherwise comments might get lost in the google doc. A google doc might be more useful when we still have very abstract ideas of what we want to do. But here we have all the use cases listed already :)

Ah, I see. I'll write the KEP, directly :) Thank you for letting me know!
Anyway, I'll wait for responses from sig-apps leads for about a week in business days.

/assign

**Comment by tenzen-y** [comment]
I have created a tracking issue on the enhancements repository. I will start writing a KEP next week.

https://github.com/kubernetes/enhancements/issues/3998

**Comment by kerthcet** [comment]
FYI: Enhancement code freeze 16th June https://www.kubernetes.dev/resources/release/

**Comment by tenzen-y** [comment]
> FYI: Enhancement code freeze at 16th June, https://www.kubernetes.dev/resources/release/

Thank you for letting me know :) I'm a beginner on k/k. So that information is much helpful!

**Comment by tenzen-y** [comment]
Sorry for the delay. I wrote the KEP: https://github.com/kubernetes/enhancements/pull/4062

**Comment by alculquicondor** [comment]
/close
in favor of https://github.com/kubernetes/enhancements/issues/3998
